### PR TITLE
Add WPGraphQL Support

### PR DESCRIPTION
**Pull request by ostowe**

### CODE REVIEW

File: inc/graphql.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Register data types and field for supporting bylines on posts.
+ */
+function register_byline_types() {
+	// Register a text profile.
+	register_graphql_object_type(
+		'TextProfile',
+		[
+			'description' => __( 'Promoted Book', 'byline-manager' ),
```
Comment (joemcgill) [inline inc/graphql.php]:
```suggestion
			'description' => __( 'A TextProfile type', 'byline-manager' ),
```

**Comment by mboynes** [comment]
If I understand this correctly, and I may not, I'm not thrilled with the DX of this implementation. As I understand it, to build the plain text byline for a post, a developer would need to query like so:

```graphql
query PostsWithBylines {
  posts {
    nodes {
      title
      byline {
        ...on Profile {
          title
        }
        ...on TextProfile {
          text
        }
      }
    }
  }
}
```

While I think that would be generally _okay_ to offer as an option, I think that should be advanced usage.

As a developer trying to add a byline to an article, I don't care about the data structures and I don't want a series of if-else clauses in my code, 99% of the time I just want a string I can render. Thinking it through, I might want the byline in the following formats (in each example, consider Jane Doe to be a `Profile` and John Smith to be a plain text entry):

* Non-html plain text (`get_the_byline()`), e.g. `"By Jane Doe and John Smith"`
* HTML text with links to their posts (`get_the_byline_posts_links()`) if applicable, e.g. `"By <a href=\"/profiles/jane-doe/\">Jane Doe</a> and John Smith"`
* HTML text with links to their website (`get_the_byline_links()`) if applicable, e.g. `"By <a href=\"https://jane.blog/\">Jane Doe</a> and John Smith"`
* Array of byline data, e.g. `[ { "text": "Jane Doe", "website": "https://jane.blog/", "postsUrl": "/profiles/jane-doe/" }, { text: "John Smith" } ]`

With that in mind, what should the field(s) look like in graphql? I see a few options. Tossing some out there: `byline` could be a field with subfields, like `byline.plainText`, `byline.postsLinks`, `byline.websiteLinks`. Maybe a separate field for the objects like `bylineEntries`, so you would query `bylineEntries.text`, `bylineEntries.website`, and `bylineEntries.postsUrl`.

**Comment by joemcgill** [comment]
@mboynes this is great feedback. Having not worked through a front end implementation yet, I was mainly thinking about providing the raw data to the front end to be as flexible as possible for the front end to handle different variations on how that data is displayed. I still think that has value, but agree that a simplified field for `byline.text` or something would be better 90% of the time, additionally we could create a shared interface between profiles that are managed as a post type and text profiles so that you don't have to query for different fields based on the profile type.

**Comment by joemcgill** [comment]
@mboynes I've updated the implementation to address your main concern, which is that most developers would want a single string value to use for bylines. I've made the `byline` field an object with two sub-fields: `bylineText` which accepts an optional argument to handle the three formatting options you described above, and a `profiles` field, which works the way the previous approach did, exposing the raw data for profiles based on their type (post data or text only). 

I still think a case could be made to try and come up with a common interface for both types to avoid the type based queries, but I think exposing the raw post data and all its mutations is a more flexible option for when someone really does want the kind of detail that would require querying the underlying data, so I left it for now.

At present, a simple query to get a specific format of the byline as a string would look like this:

```gql
query BylineQuery {
  posts {
    nodes {
      byline {
        bylineText(format: POST_LINKS)
      }
    }
  }
}
```


And a more complex query that wants to get both the bylineText and raw profile data (e.g., featured images), would look like this:

```gql
query BylineQuery {
  posts {
    nodes {
      byline {
        bylineText
        profiles {
          ... on Profile {
            title
            featuredImage {
              node {
                mediaItemUrl
              }
            }
          }
          ... on TextProfile {
            text
          }
        }
      }
    }
  }
}
```

File: inc/graphql.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+				],
+			],
+		]
+	);
+
+	/*
+	 * Register byline fields on nodes that support authors, which is also
+	 * the check byline manager uses to determine if a post supports bylines.
+	 * The byline field will return a list of either Profile posts or text
+	 * profiles based on the type stored in the post meta object
```
Comment (mboynes) [inline inc/graphql.php]: Looks like this comment is now outdated

File: inc/graphql.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+					'value' => 'links',
+				],
+				'POST_LINKS' => [
+					'value' => 'post_links',
+				],
+			],
+		]
+	);
+
+	register_graphql_object_type(
```
Comment (mboynes) [inline inc/graphql.php:70]: Since the rest of the `register_` calls have a comment, this one told me it feels left out

File: inc/graphql.php [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		'Byline',
+		[
+			'description' => __( 'Byline manager byline', 'byline-manager' ),
+			'fields'      => [
+				'bylineText' => [
+					'args'        => [
+						'format' => [
+							'type'        => 'BylineTextEnum',
+							'description' => __( 'Choose the byline format.', 'byline-manager' ),
```
Comment (mboynes) [inline inc/graphql.php:79]: The rest of the descriptions don't have end punctuation, so for the sake of consistency, I'd suggest this shouldn't either
```suggestion
							'description' => __( 'Choose the byline format', 'byline-manager' ),
```

File: inc/graphql.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+								$byline = get_the_byline_posts_links( $post->ID );
+								break;
+
+							case 'text':
+							default:
+								$byline = get_the_byline( $post->ID );
+								break;
+						}
+
+						return $byline;
```
Comment (mboynes) [inline inc/graphql.php]:
```suggestion
						switch ( $args['format'] ) {
							case 'links':
								return get_the_byline_links( $post->ID );

							case 'post_links':
								return get_the_byline_posts_links( $post->ID );
						}

						return get_the_byline( $post->ID );
```

**Review by mboynes — APPROVED** [review]
This is awesome. Once merged, can you please add a page to the wiki about using it?

File: inc/graphql.php (hunk 2 above)
Comment (joemcgill) [reply inc/graphql.php]: Indeed. The first part is still correct I'll update it.

File: inc/graphql.php (hunk 3 above)
Comment (joemcgill) [reply inc/graphql.php:70]: I feel the same way.

File: inc/graphql.php (hunk 4 above)
Comment (joemcgill) [reply inc/graphql.php:79]: Looking at the way fields are registered in the main WPGraphQL plugin, the assumption is that these should end in punctuation, so I'll update all of them.

File: inc/graphql.php (hunk 5 above)
Comment (joemcgill) [reply inc/graphql.php]: I debated this and decided to go for the more verbose version to make clear what the formats were. Can probably handle that in an inline comment just as well.

**Pull request merged**